Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -92,6 +92,62 @@ public CsharpCompiler(ErrorReporter reporter)
     }
 
     protected override void EmitBuiltInDecls(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      EmitInitNewArrays(builtIns, wr);
+
+      EmitFuncExtensions(builtIns, wr);
+    }
+
+    // Generates casts for functions of those arities present in the program, like:
+    //   public static class FuncExtensions {
+    //     public static Func<U, UResult> DowncastClone<T, TResult, U, UResult>(this Func<T, TResult> F,
+    //         Func<U, T> ArgConv, Func<TResult, UResult> ResConv) {
+    //       return arg => ResConv(F(ArgConv(arg)));
+    //     }
+    //     ...
+    //   }
+    // They aren't in any namespace to make them universally available.
+    private static void EmitFuncExtensions(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      var funcExtensions = wr.NewNamedBlock("public static class FuncExtensions");
+      foreach (var kv in builtIns.ArrowTypeDecls) {
+        int arity = kv.Key;
+
+        List<string> TyList(string prefix) {
+          var l = arity switch {
+            1 => new List<string> { prefix },
+            _ => Enumerable.Range(1, arity).Select(i => $"{prefix}{i}").ToList()
+          };
+          l.Add($"{prefix}Result");
+          return l;
+        }
+
+        string TPs(IEnumerable<string> l) => $"<{l.Comma()}>";
+
+        var us = TyList("U");
+        var ts = TyList("T");
+        var tsus = ts.Concat(us).ToList();
+
+        string ArgConv((string u, string t) tp) => $"Func<{tp.u}, {tp.t}> ArgConv";
```

Comment:
It doesn't really matter since this is not likely to be read by someone, but C# function parameters generally start with a lowerCase.

Reply:
Also, I prefer having two separate function arguments over a tuple. I'd rather deconstruct the tuple at the call-site.

Reply:
I couldn't figure out a nice way to do that in the lambda and the autogenerated names `First` and `Second` are cumbersome. Re the capitalisation: You are right, that appears to be the standard, but it feels wrong, as regular functions a capitalized.